packages/astro/src/content/runtime.ts

```diff
@@ -319,9 +334,84 @@ type RenderResult = {
 	remarkPluginFrontmatter: Record<string, any>;
 };
 
+const IMAGE_REGEX = /__ASTRO_IMAGE_="([^"]+)"/g;
+
+async function updateImageReferencesInBody(html: string, fileName: string) {
+	// @ts-expect-error Virtual module
+	const { default: imageAssetMap } = await import('astro:asset-imports');
+
+	const imageObjects = new Map<string, GetImageResult>();
+
+	// @ts-expect-error Virtual module resolved at runtime
+	const { getImage } = await import('astro:assets');
+
+	// First load all the images. This is done outside of the replaceAll
+	// function because getImage is async.
+	for (const [_full, imagePath] of html.matchAll(IMAGE_REGEX)) {
+		const decodedImagePath = JSON.parse(imagePath.replace(/&#x22;/g, '"'));
+		const id = imageSrcToImportId(decodedImagePath.src, fileName);
+
+		if (!id || imageObjects.has(id)) {
+			continue;
+		}
+
+		const imported = imageAssetMap.get(id);
+		if (!imported) {
+			continue;
+		}
+		const image: GetImageResult = await getImage({ ...decodedImagePath, src: imported });
+		imageObjects.set(imagePath, image);
+	}
+
+	return html.replaceAll(IMAGE_REGEX, (full, imagePath) => {
```

Comment:
Is there a reason why loop over `html` twice - `matchAll` and `replaceAll`?  If there's a valid reason, I believe we should document it with a comment.

Reply:
Yeah, the comment above: replaceAll can't have an async replacer, so first I get all matches and use that to asynchrounously get the image objects. I then synchronously replace them. I'm not sure of a way to do that in one step.